Add developer documentation for field-selector (#658)
flightctl/flightctl: docs/developer/architecture/field-selectors.md

`status.updated.status="UpToDate"`

- Loss of Control:
Allowing all `Status` keys to be queried restricts our ability to whitelist and only support specific keys.

#### Custom Selector for JSONB Fields

Adding Custom Selector for `status.updated.status`
```go
func (m *Device) ResolveSelector(name selector.SelectorName) (*selector.SelectorField, error) {
  if strings.EqualFold("status.updated.status", name.String()) {
		return &selector.SelectorField{
			Name:      name,
			Type:      selector.String,
			FieldName: "status.updated.status",
			FieldType: "jsonb",
		}, nil
	}
	return nil, nil
}

func (m *Device) ListSelectors() selector.SelectorNameSet {
	return selector.NewSelectorFieldNameSet().Add("status.updated.status")
}
```

> [!NOTE]
> - Mapping a selector name using `MapSelectorName` to a custom selector is also supported. This can be useful for deprecating a key.
> - When `FieldType` is defined as **JSONB**, the field-selector Adapts the `FieldName` to a valid **JSONB** key (e.g., `status -> 'updated' ->> 'status'`).
> - If the `Type` is different from **JSONB**, the field-selector will cast the key to the corresponding type, and the selector will be processed as that type.

> > - A custom selector is resolved second. It can override or hide an existing resolver defined by the resource model.

Reviewer: I think it would be best to point about the resolution order more explicitly.
```suggestion
> - A custom selector is resolved after mapped selectors. <Rest here>
```

## Kubernetes Selector Package

The field-selector uses the Kubernetes selector to initially parse the query.

The package was added in `/pkg/k8s/selector`, and the following modifications were made:

- Kubernetes label selectors are now used for both fields and labels.
- The functions `validateLabelKey` and `validateLabelValue` are applied only to labels.
- Fields have a modified lexer to support escaping and RHS symbols (e.g., `key in (x=y)`).

> - A new operators, `contains` and `notcontains`, were added.

Reviewer: ```suggestion
- Two new operators, `contains` and `notcontains`, were added.
```
